File: server/src/routes/small-order-fee-settings.js
const router = require('express').Router();
const db = require('../db');
const asyncHandler = require('../asyncHandler');
const { pickFields, requireRole } = require('../utils');

const FIELDS = ['is_active', 'threshold_amount', 'fee_amount'];

/** Chỉ giữ 1 dòng đang áp dụng — dòng mới nhất theo updated_at, xem
 * hofa-db/108_buy_on_behalf_price_fold_and_small_order_fee.sql. */
async function currentSettings() {
  return db.queryOne('SELECT * FROM small_order_fee_settings ORDER BY updated_at DESC LIMIT 1');
}

// Công khai (không cần đăng nhập) — admin app cần đọc trước khi đăng nhập xong provider mới
// chạy được, giữ công khai cho nhất quán với các file *-settings.js khác.
router.get('/small-order-fee-settings', asyncHandler(async (req, res) => {
  const row = await currentSettings();
  res.json({ ok: true, data: row });
}));

router.patch('/small-order-fee-settings', asyncHandler(async (req, res) => {
  requireRole(req.ctx, ['admin']);
  const existing = await currentSettings();
  const data = { ...pickFields(req.body, FIELDS), updated_at: new Date().toISOString(), updated_by: req.ctx.userId };
  const updated = existing
    ? await db.updateById('small_order_fee_settings', existing.id, data)
    : await db.insertRow('small_order_fee_settings', data);
  res.json({ ok: true, data: updated });
}));

module.exports = router;

File: hofa_admin_app/lib/models/small_order_fee_settings.dart
/// Phí cộng thêm cho đơn giá trị nhỏ — áp dụng MỌI cửa hàng (không riêng mua hộ). Đơn có
/// subtotal (đã gồm % phí mua hộ nếu có) dưới [thresholdAmount] thì total_amount cộng thêm
/// [feeAmount]. Xem hofa-db/108_buy_on_behalf_price_fold_and_small_order_fee.sql.
class SmallOrderFeeSettings {
  final String? id;
  final bool isActive;
  final int thresholdAmount;
  final int feeAmount;

  SmallOrderFeeSettings({
    this.id,
    this.isActive = true,
    required this.thresholdAmount,
    required this.feeAmount,
  });

  factory SmallOrderFeeSettings.fromJson(Map<String, dynamic> json) =>
      SmallOrderFeeSettings(
        id: json['id'] as String?,
        isActive: json['is_active'] as bool? ?? true,
        thresholdAmount: (json['threshold_amount'] as num?)?.toInt() ?? 30000,
        feeAmount: (json['fee_amount'] as num?)?.toInt() ?? 5000,
      );

  /// Mặc định dùng khi server chưa có dòng cấu hình nào (chưa từng chạy migration).
  factory SmallOrderFeeSettings.fallback() =>
      SmallOrderFeeSettings(thresholdAmount: 30000, feeAmount: 5000);

  Map<String, dynamic> toJson() => {
    'is_active': isActive,
    'threshold_amount': thresholdAmount,
    'fee_amount': feeAmount,
  };
}
